R/tag_query.R

```diff
@@ -0,0 +1,1601 @@
+
+# Use for `has()` functionality
+envirMap <- function() {
+  map <- fastmap()
+  list(
+    keys = map$keys,
+    asList = function() {
+      unname(map$as_list())
+    },
+    has = function(envir) {
+      map$has(envir$envKey)
+    },
+    set = function(envir, value = envir) {
+      map$set(envir$envKey, value)
+    },
+    remove = function(envir) {
+      map$remove(envir$envKey)
+    }
+  )
+}
+# Use for consistent `asList()` order
+envirStack <- function() {
+  stack <- faststack()
+  list(
+    push = stack$push,
+    asList = stack$as_list,
+    uniqueList = function() {
+      unique(stack$as_list())
+    }
+  )
+}
+
+# (Used for `unique_envirStack()` only. Do not use directly!)
+envirStackUni_ <- function() {
+  map <- fastmap()
+  stack <- faststack()
+  list(
+    push = function(env) {
+      key <- env$envKey
+      if (!map$has(key)) {
+        # mark the key as _seen_
+        map$set(key, TRUE)
+        # add the env
+        stack$push(env)
+      }
+    },
+    uniqueList = stack$as_list
+  )
+}
+# Use to retrieve unique environments (eg: `tq$parent()`)
+envirStackUnique <- function() {
+  stack <- envirStack()
+  count <- 0
+  list(
+    push = function(env) {
+      count <<- count + 1
+      if (count == 500) {
+        # convert the current stack to a `envirStackUni_()`
+        newStack <- envirStackUni_()
+        walk(stack$asList(), newStack$push)
+        stack <<- newStack
+      }
+      stack$push(env)
+    },
+    uniqueList = function() {stack$uniqueList()}
+  )
+}
+
+
+
+
+
+
+# Retrieve all attributes that can be manually set
+# ?attr
+# Note that some attributes (namely ‘class’, ‘comment’, ‘dim’,
+# ‘dimnames’, ‘names’, ‘row.names’ and ‘tsp’) are treated specially
+# and have restrictions on the values which can be set.
+safeAttrValues <- function(x) {
+  badElAttrs <- c("class", "comment", "dim", "dimnames", "names", "row.names", "tsp")
+  attrVals <- attributes(x)
+  attrVals[badElAttrs] <- NULL
+  attrVals
+}
+
+# Convert a list to an environment and keep class and attribute information
+safeListToEnv <- function(x, newClass = NULL) {
+  xList <- x
+
+  ret <- list2env(xList, new.env(parent = emptyenv()))
+
+  attrVals <- safeAttrValues(xList)
+  walk2(names(attrVals), attrVals, function(attrName, attrValue) {
+    attr(ret, attrName) <<- attrValue
+  })
+
+  class(ret) <- c(newClass, class(xList), "environment")
+  ret
+}
+
+# Convert an environment to a list and keep class and attribute information
+safeEnvToList <- function(x, oldClass = NULL) {
+  xEnv <- x
+  ret <- as.list.environment(xEnv, all.names = TRUE)
+
+  attrVals <- safeAttrValues(xEnv)
+  walk2(names(attrVals), attrVals, function(attrName, attrValue) {
+    attr(ret, attrName) <<- attrValue
+  })
+
+  class(ret) <- setdiff(class(x), c(oldClass, "environment"))
+  ret
+}
+
+
+# Do not export to encourage direct use of `tagQuery()`
+asTagEnv <- function(x) {
+  if (isTagQuery(x)) {
+    return(asTagEnv(x$root()))
```

Comment:
`x$root()` will always be a tag env, right? If so, no need for `asTagEnv()`.

Reply:
And now that I think of it, does this even need to accept tagQuery objects? If we don't actually use that anywhere, it would be conceptually simpler to just throw an error if a tagQuery object is passed in.